### ISSUE TITLE
🐛 Workspace creation errors leave workspace in a limbo state 

### ISSUE THREAD

**Issue by KamalAman** [hashicorp/terraform-cloud-operator#214]
### Operator Version, Kind and Kubernetes Version
 - Operator version: 2.0.0-beta7
 - Kind: Workspace
 - Kubernetes version: 1.24

### Description
Workspace creation errors leave workspace in a limbo state. For example add a space in a tag value (which is invalid) and apply the Workspace Resource. The TFC  workspace will be created, but it will error during creation, and not properly linked to the workspace. 

### Output Log

```
2023-07-08T02:00:55Z	INFO	Reconcile Tags	{"workspace": {"name":"tenant-09855479-5ff3-4fd1-a4bd-531aff5dfb62-c8f51a80","namespace":"tf-tenants"}, "msg": "adding tags from the workspace"}
2023-07-08T02:00:55Z	ERROR	Reconcile Tags	{"workspace": {"name":"tenant-09855479-5ff3-4fd1-a4bd-531aff5dfb62-c8f51a80","namespace":"tf-tenants"}, "msg": "reconcile tags", "error": "invalid attribute\n\nTags must be one or more characters; can include letters, numbers, colons, hyphens, and underscores; and must begin and end with a letter or number"}
2023-07-08T02:00:55Z	ERROR	Workspace Controller	{"workspace": {"name":"tenant-09855479-5ff3-4fd1-a4bd-531aff5dfb62-c8f51a80","namespace":"tf-tenants"}, "msg": "reconcile workspace", "error": "invalid attribute\n\nTags must be one or more characters; can include letters, numbers, colons, hyphens, and underscores; and must begin and end with a letter or number"}
2023-07-08T02:01:10Z	INFO	Workspace Controller	{"workspace": {"name":"tenant-09855479-5ff3-4fd1-a4bd-531aff5dfb62-c8f51a80","namespace":"tf-tenants"}, "msg": "new reconciliation event"}
2023-07-08T02:01:10Z	INFO	Spec Validation	{"workspace": {"name":"tenant-09855479-5ff3-4fd1-a4bd-531aff5dfb62-c8f51a80","namespace":"tf-tenants"}, "msg": "validating instance object spec"}
2023-07-08T02:01:10Z	INFO	Spec Validation	{"workspace": {"name":"tenant-09855479-5ff3-4fd1-a4bd-531aff5dfb62-c8f51a80","namespace":"tf-tenants"}, "msg": "spec is valid"}
2023-07-08T02:01:10Z	INFO	Reconcile Workspace	{"workspace": {"name":"tenant-09855479-5ff3-4fd1-a4bd-531aff5dfb62-c8f51a80","namespace":"tf-tenants"}, "msg": "reconciling workspace"}
2023-07-08T02:01:10Z	INFO	Reconcile Workspace	{"workspace": {"name":"tenant-09855479-5ff3-4fd1-a4bd-531aff5dfb62-c8f51a80","namespace":"tf-tenants"}, "msg": "status.WorkspaceID is empty, creating a new workspace"}
2023-07-08T02:01:10Z	ERROR	Reconcile Workspace	{"workspace": {"name":"tenant-09855479-5ff3-4fd1-a4bd-531aff5dfb62-c8f51a80","namespace":"tf-tenants"}, "msg": "failed to create a new workspace", "error": "invalid attribute\n\nName has already been taken"}
2023-07-08T02:01:10Z	ERROR	Reconcile Workspace	{"workspace": {"name":"tenant-09855479-5ff3-4fd1-a4bd-531aff5dfb62-c8f51a80","namespace":"tf-tenants"}, "msg": "failed to create a new workspace", "error": "invalid attribute\n\nName has already been taken"}
```

### Expected Behavior
Since the TFC Workspace was created, the Workspace ID should be linked to the Workspace Resource

### Actual Behavior
The Workspace ID is not linked to the Workspace Resource, so the operator tries to create another Workspace resulting in the  `Name has already been taken` error. 

### Additional Context
The Workspace being unlinked causes cleanup and reconciliation issues. The only way to resolve this is to manually go to TFC and remove the Workspace.

### Community Note

* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request.
* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.

**Comment by arybolovlev** [comment]
Hi @KamalAman,

Thank you for raising this. I can see two issues here:

- the validator doesn't handle the case when tag is invalid.
- the workspace controller doesn't update state when an error happens during the reconciliation and thus in the case when a workspace is created or recreated it may cause and observed issue.

We need to update the validator to decline invalid CR earlier and add a defer function when workspace reconciliation ends with an error.

Thanks!

**Comment by KamalAman** [comment]
While I agree better validation would be better, i have actually ran into this issue a number of times over the last week. 

For example i experienced a  failed notification endpoint test which caused an error. I don't think there is any sort of validation you could do to detect that one.

I think no matter how well the validation does its job, there will be things outside of its scope that would cause a workspace errors. I think it is important, from a reliability standpoint of view, to change the error handling logic to update the state with the workspaceId since the workspace was created. Otherwise manual intervention both in TFC and the Cluster a required.

**Comment by gautambaghel** [comment]
This happens with the AgentPool CRD as well

**Comment by arybolovlev** [comment]
Hi @KamalAman,

The fix will be available in the upcoming release `2.0.0-beta8`. Thank you for reporting this issue!

@gautambaghel, please open a new issue. Thanks!

**Comment by KamalAman** [comment]
Thanks, the new version works well can now I can device the reason why a workspace fails e.g. special characters in the tag
